README.md
# svg-icon.vue

[![npm version](https://img.shields.io/npm/v/@yaohaixiao/svg-icon.vue)](https://www.npmjs.com/package/@yaohaixiao/svg-icon.vue)
[![prettier code style](https://img.shields.io/badge/code_style-prettier-ff69b4.svg)](https://prettier.io)
[![Coverage](https://codecov.io/gh/yaohaixiao/svg-icon.vue/branch/main/graph/badge.svg)](https://codecov.io/gh/yaohaixiao/vg-icon.vue)
[![vue2](https://img.shields.io/badge/vue-2.x-brightgreen.svg)](https://vuejs.org/)
[![npm downloads](https://img.shields.io/npm/dm/@yaohaixiao/svg-icon.vue)](https://npmcharts.com/compare/@yaohaixiao/svg-icon.vue?minimal=true)
[![MIT License](https://img.shields.io/github/license/yaohaixiao/svg-icon.vue.svg)](https://github.com/yaohaixiao/svg-icon.vue/blob/master/LICENSE)

svg-icon.vue 为开发者提供了一个用来在 VUE 项目中显示 SVG 图标的工具。


## Install

svg-icon.vue 的安装十分简单，可以直接在命令行中运行 npm install 命令安装，也可以直接在项目的 package.js 中配置依赖：

### Install from the command line

```shell
npm i @yaohaixiao/svg-icon.vue
```

### Install via package.json:

```shell
"@yaohaixiao/svg-icon.vue": "2.x.x"
```


## API Documentation

### Online docs

Online docs：[https://yaohaixiao.github.io/svg-icon.vue/#/docs](https://yaohaixiao.github.io/svg-icon.vue/#/docs)

### Local docs

```shell
# serve with hot reload at localhost:48081
npm run start
```

### Methods

svg-icon.vue 的 API 对外提供了 render() 和 getSymbols() 两个方法，以下是这两个方法的具体使用细节。

#### render()

SvgIcon 组件至 2.0.0 版本开始，不再直接绘制 default 图标集。需要手动导入 render() 方法。

```js
import { render } from 'svg-icon.vue/utils/utils'
```

##### Grammar

```js
render(iconSet)
```

##### Options

| Name              | Type           | Default   | Description                  |
|-------------------|----------------|-----------|------------------------------|
| `iconSet`         | `Object,Array` | --        | 可选，图标集数据对象或者图标集数组。           |
| `iconSet.title`   | `String`       | --        | 必选，图标集名称。                    |
| `iconSet.symbols` | `Array`        | --        | 必选，存储 svg 图标的 symbols 数组 图标集 |

绘制 default 图标集：

```js
import { render } from 'svg-icon.vue/utils/utils'

render()
```

绘制自定义的图标集：

```js
// svg-icon.vue 内置了多个免费且设计优美的图标集：
// 01. svg-icon.vue/assets/brankic-1979
// 02. svg-icon.vue/assets/broccolidry
// 03. svg-icon.vue/assets/default
// 04. svg-icon.vue/assets/eighty-shades
// 05. svg-icon.vue/assets/entypo
// 06. svg-icon.vue/assets/feather
// 07. svg-icon.vue/assets/font-awesome
// 08. svg-icon.vue/assets/hawcons
// 09. svg-icon.vue/assets/iconic
// 10. svg-icon.vue/assets/linecons
// 11. svg-icon.vue/assets/material
// 12. svg-icon.vue/assets/meteocons
// 13. svg-icon.vue/assets/steadysets
// 14. svg-icon.vue/assets/typicons
// 15. svg-icon.vue/assets/vicons
// 16. svg-icon.vue/assets/wpzoom
// 17. svg-icon.vue/assets/zondicons
import defaultSet from 'svg-icon.vue/assets/default'
import fontAwesomeSet from 'svg-icon.vue/assets/font-awesome'

import { render, getSymbols } from 'svg-icon.vue/utils/utils'

// 绘制单个图标集
render(fontAwesomeSet)

// 一次绘制多个图标集
render([defaultSet, fontAwesomeSet])
```

说明：如果你只想使用内置图标集中特定的几个图标，你可以在 API 文档的 [icons](https://yaohaixiao.github.io/svg-icon.vue/#/icons) 相关页面选中需要的图标，然后点击图标购物车，我们会自动为你生成自定义图标集。并且提供下载 SVG 图标集和 JS 源代码复制功能。 如果你仅仅想使用某个图标，我们也同时提供单独下载 SVG 图标的功能。

#### getSymbols()

getSymbols() 方法返回使用 render() 方法绘制的图标集中所有 symbols （数组）数据。

```js
import { render } from 'svg-icon.vue/utils/utils'

render()

const symbols = getSymbols()
// => 返回 default 图标集中的 symbols （数组）数据
```

### Component

svg-icon.vue 为开发者提供了 SvgIcon 组件，使用它来显示 render() 方法绘制的 svg 图标库中某个图标。

#### Import package

使用 render() 方法绘制 svg 图标后，就可以调用 SvgIcon 组件显示图标了。

```js
// 导入SvgIcon组件
import SvgIcon from 'svg-icon.vue'

export default {
  components: {
    SvgIcon
  }
}
```

#### SvgIcon 组件

SvgIcon.vue 是 svg-icon.vue 的核心组件，用来显示 render() 方法绘制的 svg 图标集中的图标。

##### Grammar

SvgIcon 组件使用起来十分简单，调用语法如下：

```vue
<svg-icon name="help" size="16" color="#999999" />
```

##### Options

| Name    | Type                   | Default   | Description                                                                              |
|---------|------------------------|-----------|------------------------------------------------------------------------------------------|
| `name`  | `String`               | --        | 必选，图标名称：图标库中所有图标的名称。  |
| `size`  | `Number, String, Array`| 16        | 可选，用来指定 icon 图标大小。number 和 string 类型时，可选值为 0 以上的整数, 宽度和高度值相等；array时数组长度为 2，分别代表宽度和高度；                                                                       |
| `color` | `String`               | --        | 可选，用来指定 icon 图标颜色。                                                                       |


## Usage

SvgIcon 组件的使用方法如下：

### Basic

size 和 color 参数是可选的，只用设置 name 属性（ icon 的名称）即可，默认 size 值为 16px 大小。当然，也可以自定义设置 size 大小。

```vue
<template>
  <svg-icon name="plugins" />
  <svg-icon name="database-read" size="24" />
  <svg-icon name="database-lock" :size="32" />
  <svg-icon name="deploy" :size="[44, 44]" />
</template>

<script>
import { render } from 'svg-icon.vue/utils/utils'
import SvgIcon from 'svg-icon.vue'

render()

export default {
  components: {
    SvgIcon
  }
}
</script>
```

### Set color

SvgIcon 组件即可以使用包含 icon 的父节点设置 CSS 样式控制图标颜色，也可以直接设置 color 属性设置颜色。

#### Set color with CSS

通过设置 SvgIcon 父组件的 CSS 颜色，SvgIcon 组件的图标颜色将其父元素 case-icon 的 color 值一致。

```vue
<template>
  <span class="case-icon">
    <svg-icon name="database-read" :size="24" />
  </span>
  <span class="case-icon">
    <svg-icon name="database-lock" :size="24" />
  </span>
  <span class="case-icon">
    <svg-icon name="plugins" :size="24" />
  </span>
</template>

<script>
import { render } from 'svg-icon.vue/utils/utils'
import SvgIcon from 'svg-icon.vue'

render()

export default {
  components: {
    SvgIcon
  }
}
</script>

<style scoped lang="less">
  .case-icon {
    display: inline-block;
    margin: 0 8px;
    color: @third_text_color;
    vertical-align: middle;
    cursor:pointer;
    overflow: hidden;

    &:hover {
      color: @primary_color;
    }
  }
</style>
```

#### Set color with 'color' attribute

设置 color 属性后，SvgIcon 的父组件通过 CSS 设置的 color 样式将不再起作用。

```vue
<template>
  <span class="case-icon">
    <svg-icon name="aside-environment" :size="24" color="#07B759"/>
  </span>
  <span class="case-icon">
    <svg-icon name="aside-branch" :size="24" color="#507AFE"/>
  </span>
  <span class="case-icon">
    <svg-icon name="aside-pc" :size="24" color="#FF9901"/>
  </span>
</template>


<script>
import { render } from 'svg-icon.vue/utils/utils'
import SvgIcon from 'svg-icon.vue'

render()

export default {
  components: {
    SvgIcon
  }
}
</script>

<style scoped lang="less">
.case-icon {
  display: inline-block;
  margin: 0 8px;
  color: @third_text_color;
  vertical-align: middle;
  cursor:pointer;
  overflow: hidden;

  &:hover {
    color: @primary_color;
  }
}
</style>
```

### Use Customized Icon Set

```js
/**
 * ico-moon.js - icoMoon 图标集（图标集实例）
 * =============================================================
 * Created By: Yaohaixiao
 * Update: 2022.10.16
 */
const icoMoonSet = {
  title: 'icoMoon 图标集',
  symbols: [
    '<symbol id="icon-moon-home" viewBox="0 0 16 16"><path d="M16 9.226l-8-6.21-8 6.21v-2.532l8-6.21 8 6.21zM14 9v6h-4v-4h-4v4h-4v-6l6-4.5z"></path></symbol>',
    '<symbol id="icon-moon-home2" viewBox="0 0 16 16"><path d="M8 0.5l-8 8 1.5 1.5 1.5-1.5v6.5h4v-3h2v3h4v-6.5l1.5 1.5 1.5-1.5-8-8zM8 7c-0.552 0-1-0.448-1-1s0.448-1 1-1c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path></symbol>',
    '<symbol id="icon-moon-home3" viewBox="0 0 16 16"><path d="M16 9.5l-3-3v-4.5h-2v2.5l-3-3-8 8v0.5h2v5h5v-3h2v3h5v-5h2z"></path></symbol>'
  ]
}

export default icoMoonSet
```

```vue
<template>
  <span class="case-icon">
    <svg-icon name="moon-home" :size="24"/>
  </span>
  <span class="case-icon">
    <svg-icon name="moon-home2" :size="24"/>
  </span>
  <span class="case-icon">
    <svg-icon name="moon-home3" :size="24"/>
  </span>
</template>

<script>
import SvgIcon from 'svg-icon.vue'

import { render } from 'svg-icon.vue/utils/utils'
// 自定义图标集可以是从外部 .js 文件导入的模块
// 也可以是类似 icoMoonSet 格式的 JavaScript 对象数据
import icoMoonSet from 'assets/ico-moon'

render(icoMoonSet)

export default {
  components: {
    SvgIcon
  }
}
</script>
```


## Icons

SvgIcon 组件已经内置了一些常用图标：[https://yaohaixiao.github.io/svg-icon.vue/#/icons](https://yaohaixiao.github.io/svg-icon.vue/#/icons)


## License

JavaScript Code Licensed under [MIT License](http://opensource.org/licenses/mit-license.html).

API Documentation Licensed under [CC BY 3.0](http://creativecommons.org/licenses/by/3.0/)


CHANGELOG.MD
# [2.7.0](https://github.com/yaohaixiao/svg-icon.vue/compare/2.6.1...2.7.0) (2022-12-26)


### Features

* 利用批量导入，添加 EverIcons 图标集 ([e0f9dd1](https://github.com/yaohaixiao/svg-icon.vue/commit/e0f9dd16c7544b1425dfe353e9fe8a16d59d2c41))



## [2.6.1](https://github.com/yaohaixiao/svg-icon.vue/compare/2.6.0...2.6.1) (2022-12-26)


### Bug Fixes

* 修复部分图标集页面搜索报错问题 ([f22c3f9](https://github.com/yaohaixiao/svg-icon.vue/commit/f22c3f94eca6b7dbc63acded94d076234be6f4c8))
* 修复更新绘制图标时，重复绘制之前已经绘制的图标问题 ([dd0a5bb](https://github.com/yaohaixiao/svg-icon.vue/commit/dd0a5bb9dd7f74f07d8612efd1f02b24a7b800b7))
* 修复使用分页的界面，清空搜索没有进行分页处理的问题 ([cd3a9b1](https://github.com/yaohaixiao/svg-icon.vue/commit/cd3a9b13c45d9649441ed80733c23c6137431cb5))



# [2.6.0](https://github.com/yaohaixiao/svg-icon.vue/compare/2.5.1...2.6.0) (2022-11-23)


### Features

* 为每个图标集添加 icons.js 用于直接绘制该图标库的所有 SVG 图标； ([69d53af](https://github.com/yaohaixiao/svg-icon.vue/commit/69d53af3b404dae1263fee65576ec5c417e4d89e))



## [2.5.1](https://github.com/yaohaixiao/svg-icon.vue/compare/2.5.0...2.5.1) (2022-11-22)


### Bug Fixes

* 修复发布 package 配置报错问题 ([57771ca](https://github.com/yaohaixiao/svg-icon.vue/commit/57771ca488f24a84632344d9ee3ddcd13deabe25))



# [2.5.0](https://github.com/yaohaixiao/svg-icon.vue/compare/2.4.1...2.5.0) (2022-11-21)


### Features

* 添加绘制所有内置SVG图标的 assets/index.js 模块 ([62f7433](https://github.com/yaohaixiao/svg-icon.vue/commit/62f7433937b43d6e8042e7cddc7b417475fde007))



## [2.4.1](https://github.com/yaohaixiao/svg-icon.vue/compare/2.4.0...2.4.1) (2022-11-18)


### Bug Fixes

* 调整默认图标库 boldSet 图标集的名称，移除重复的图标 ([36a2506](https://github.com/yaohaixiao/svg-icon.vue/commit/36a2506512a12392d85ad4e10c2e1306ac57f971))



# [2.4.0](https://github.com/yaohaixiao/svg-icon.vue/compare/2.3.1...2.4.0) (2022-11-13)


### Features

* 添加 getSymbols() 方法 ([f3dad77](https://github.com/yaohaixiao/svg-icon.vue/commit/f3dad77293b9eeb63b7a5c1931fc87e799b609f8))



## [2.3.1](https://github.com/yaohaixiao/svg-icon.vue/compare/2.3.0...2.3.1) (2022-11-11)


### Bug Fixes

* 修复 render 绘制多个图标集的问题，调整 docs 中的公共组件 ([4ad44ac](https://github.com/yaohaixiao/svg-icon.vue/commit/4ad44ac177f114cf3e3f9322bda93838d842dc45))



# [2.3.0](https://github.com/yaohaixiao/svg-icon.vue/compare/2.2.1...2.3.0) (2022-11-10)


### Features

* render() 方法支持一次绘制多个图标集，并生成对应的 API 文档 ([ecbb627](https://github.com/yaohaixiao/svg-icon.vue/commit/ecbb627611800e3c7ffe1e20b8a19d44802b5ba3))



## [2.2.1](https://github.com/yaohaixiao/svg-icon.vue/compare/2.2.0...2.2.1) (2022-11-10)


### Bug Fixes

* 修复 nav-electrocardiogram 图标 symbol 标签 SVG 格式错误 ([61c4510](https://github.com/yaohaixiao/svg-icon.vue/commit/61c45100f91a1410ee5330a562d5728f0fe1a539))



# [2.2.0](https://github.com/yaohaixiao/svg-icon.vue/compare/2.1.0...2.2.0) (2022-11-09)


### Features

* 为 default 图标库添加新的图标 ([8972f44](https://github.com/yaohaixiao/svg-icon.vue/commit/8972f448a6a07e2f63164030be098feaa78a34c2))



# [2.1.0](https://github.com/yaohaixiao/svg-icon.vue/compare/2.0.0...2.1.0) (2022-11-08)


### Features

* 添加图标集和单个图标下载功能 ([87bc5c0](https://github.com/yaohaixiao/svg-icon.vue/commit/87bc5c0f9fee39b8ede8f324140cef2ea8b82ab5))



# [1.4.0](https://github.com/yaohaixiao/svg-icon.vue/compare/1.3.0...1.4.0) (2022-10-16)


### Features

* 1. 添加新的实例；2.添加调用自定义图标功能；3. 重新成成 docs 文档； ([82afe18](https://github.com/yaohaixiao/svg-icon.vue/commit/82afe18e0f65f5d8826481bb4e8252354267c52c))


# 1.3.0 (2022-10-15)

svg-icon.vue 为开发者提供了用来在 VUE 项目中显示 SVG 图标的工具。



vue.config.js
/**
 * vue.config.js - vue 工程配置
 * =============================================================
 * Created By: Yaohaixiao
 * Update: 2022.10.8
 */
const path = require('path')
const BundleAnalyzerPlugin =
  require('webpack-bundle-analyzer').BundleAnalyzerPlugin
const PreloadWebpackPlugin = require('@vue/preload-webpack-plugin')
const HtmlWebpackPlugin = require('html-webpack-plugin')
const HtmlInlineScriptPlugin = require('html-inline-script-webpack-plugin')
const HtmlWebpackInlineSourcePlugin = require('html-webpack-inline-source-plugin')
const pkg = require('./package.json')

const trim = (str) => {
  if (str.trim) {
    return str.trim()
  } else {
    return str.replace(/^\s+/g, '').replace(/\s+$/g, '')
  }
}

const resolve = (dir) => {
  return path.join(__dirname, dir)
}

module.exports = {
  publicPath: './',
  productionSourceMap: false,
  lintOnSave: process.env.NODE_ENV === 'development',
  devServer: {
    host: 'localhost',
    port: 48081,
    hot: true,
    open: true
    // historyApiFallback: {
    //   index: '/index.html'
    // }
  },
  pluginOptions: {
    'style-resources-loader': {
      preProcessor: 'less',
      patterns: [path.resolve(__dirname, './documentation/theme/index.less')]
    }
  },
  configureWebpack: {
    plugins: [
      new HtmlInlineScriptPlugin({
        scriptMatchPattern: [/runtime[.-]?(.*?)\.js$/, /app[.-]?(.*?)\.js$/]
      }),
      new HtmlWebpackInlineSourcePlugin(HtmlWebpackPlugin)
    ],
    resolve: {
      alias: {
        '@': resolve('src'),
        vue$: 'vue/dist/vue.esm.js',
        $router: resolve('./documentation/router'),
        $components: resolve('./documentation/components'),
        $assets: resolve('./documentation/assets'),
        $utils: resolve('./documentation/utils'),
        $mixins: resolve('./documentation/mixins'),
        $pages: resolve('./documentation/pages'),
        $views: resolve('./documentation/views')
      }
    },
    module: {
      rules: [
        {
          test: /\.(htm|md)(\?.*)?$/,
          loader: 'raw-loader'
        }
      ]
    },
    mode: 'production',
    optimization: {
      usedExports: true
    }
  },
  chainWebpack(config) {
    let buildFor = process.env.BUILD_FOR

    // when process.env.npm_config_report is true, build analyzer
    // 打包分析配置
    config.when(process.env.npm_config_report, (config) => {
      config.plugin('bundle-analyzer').use(BundleAnalyzerPlugin).end()
    })

    if (buildFor) {
      buildFor = trim(buildFor)
    } else {
      buildFor = ''
    }

    // 构建 API 文档的 HTMLWebpackPlugin 插件配置
    config.plugin('html').tap((args) => {
      const description = `${pkg.description}`
      // 将脚本插入到 body 标签中
      args[0].inject = 'body'
      // 使用 HtmlWebpackInlineSourcePlugin 插件将
      // app.css 公共样式写入到 index.html，以优化性能
      args[0].inlineSource = 'app.(.*?).(css)$'
      args[0].title = `svg-icon.vue - v${pkg.version} | ${description}`
      args[0].keywords = `javascript,svg,icon,svg-icon.vue,vue,vue.js`
      args[0].description = description

      return args
    })

    config.when(buildFor === 'docs', (config) => {
      // http://www.yaohaixiao.com/blog/preload-key-requests/
      // it can improve the speed of the first screen, it is recommended to turn on preload
      // 预加载资源
      config
        .plugin('preload')
        .use(PreloadWebpackPlugin)
        .tap(() => [
          {
            rel: 'preload',
            // https://github.com/vuejs/vue-cli/blob/dev/packages/@vue/cli-service/lib/config/app.js#L171
            // 预加载忽略以下文件：
            // .map 文件是非必须的
            // hot-update 为本地开发时的临时文件
            // runtime.js 和 app.js 这些公共脚本使用 HtmlInlineScriptPlugin 插件写入到 index.html 文件中
            fileBlacklist: [
              /\.map$/,
              /hot-update\.js$/,
              /runtime[.-]?(.*?)\.js$/,
              /app[.-]?(.*?)\.(js|css)$/
            ],
            // initial, asyncChunks, all, allAssets
            include: 'initial'
          }
        ])

      // http://www.yaohaixiao.com/blog/preload-key-requests/
      // prefetch 预取配置，优化前端性能
      config
        .plugin('prefetch')
        .use(PreloadWebpackPlugin)
        .tap(() => [
          {
            rel: 'prefetch',
            // 过滤掉重复的公共样式
            fileBlacklist: [/(Page|Module)(.*?)\.(js|css)$/],
            include: {
              type: 'asyncChunks',
              entries: ['app']
            }
          }
        ])

      // https://webpack.js.org/configuration/optimization/#optimizationruntimechunk
      config.optimization.runtimeChunk('single').splitChunks({
        // 表示选择哪些 chunks 进行分割，可选值有：
        // async
        // initial
        // all
        chunks: 'all',

        // 表示新分离出的chunk必须大于等于minSize，默认为30000，约30kb。
        // minSize: 30000,

        // 表示一个模块至少应被minChunks个chunk所包含才能分割。默认为1。
        // minChunks: 2,

        // 表示按需加载文件时，并行请求的最大数目。默认为5。
        // maxAsyncRequests: 5,

        // 表示加载入口文件时，并行请求的最大数目。默认为3。
        // maxInitialRequests: 3,

        // 表示拆分出的chunk的名称连接符。默认为~。如chunk~vendors.js
        // automaticNameDelimiter: '~',

        // 设置chunk的文件名。默认为true。当为true时，splitChunks基于chunk和cacheGroups的key自动命名。
        // name: true,

        // cacheGroups 下可以配置多个组，每个组根据test设置条件，符合test条件的模块，就分配到该组。
        // 模块可以被多个组引用，但最终会根据priority来决定打包到哪个组中。默认将所有来自 node_modules
        // 目录的模块打包至vendors组，将两个以上的chunk所共享的模块打包至default组。
        cacheGroups: {
          vue: {
            name: 'chunk-vuejs',
            test: /[\\/]node_modules[\\/]_?vue(.*)/,
            priority: 30,
            chunks: 'initial',
            reuseExistingChunk: true
          },
          libs: {
            name: 'chunk-libs',
            test: /[\\/]node_modules[\\/]/,
            priority: 26,
            reuseExistingChunk: true
          },
          commons: {
            name: 'chunk-commons',
            test: resolve('documentation/components'),
            priority: 26,
            reuseExistingChunk: true
          }
        }
      })
    })
  },

  transpileDependencies: true
}
